daemon/api_themes.go

```diff
@@ -0,0 +1,245 @@
+		Names: []string{"content"},
+		Slots: true,
+	})
+	for _, info := range infos {
+		for _, slot := range info.Slots {
+			var content string
+			// The content interface ensures this attribute exists
+			if err := slot.Attr("content", &content); err != nil {
+				return nil, nil, nil, err
+			}
+			var themes *[]string
+			switch content {
+			case "gtk-3-themes":
+				themes = &gtkThemes
+			case "icon-themes":
+				themes = &iconThemes
+			case "sound-themes":
+				themes = &soundThemes
+			default:
+				continue
+			}
+			var sources []interface{}
+			if err := slot.Attr("source.read", &sources); err != nil {
+				continue
+			}
+			for _, s := range sources {
+				if path, ok := s.(string); ok {
+					*themes = append(*themes, filepath.Base(path))
+				}
+			}
+		}
+	}
+	sort.Strings(gtkThemes)
+	sort.Strings(iconThemes)
+	sort.Strings(soundThemes)
+	return gtkThemes, iconThemes, soundThemes, nil
+}
+
+var badPkgCharRegexp = regexp.MustCompile(`[^a-z]+`)
+
+func themePackageCandidates(prefix, themeName string) []string {
+	themeName = strings.ToLower(themeName)
+	themeName = badPkgCharRegexp.ReplaceAllString(themeName, "-")
+	themeName = strings.Trim(themeName, "-")
+
+	var packages []string
+	for themeName != "" {
+		packages = append(packages, prefix+themeName)
+		pos := strings.LastIndexByte(themeName, '-')
+		if pos < 0 {
+			break
+		}
+		themeName = themeName[:pos]
+	}
+	return packages
+}
+
+func getThemeStatusForType(ctx context.Context, theStore snapstate.StoreService, user *auth.UserState, prefix string, themes, installed []string, toInstall map[string]bool) (map[string]themeStatus, error) {
+	status := make(map[string]themeStatus, len(themes))
+
+	for _, theme := range themes {
+		// Skip duplicates
+		if _, ok := status[theme]; ok {
+			continue
+		}
+		if strutil.SortedListContains(installed, theme) {
+			status[theme] = themeInstalled
+			continue
+		}
+		status[theme] = themeUnavailable
+		for _, name := range themePackageCandidates(prefix, theme) {
+			if info, err := theStore.SnapInfo(ctx, store.SnapSpec{Name: name}, user); err == nil {
+				// Only mark the theme as available if
+				// it has been published to the stable
+				// channel.
+				if info.Channel == "stable" {
+					status[theme] = themeAvailable
+					toInstall[name] = true
+				}
+				break
+			} else if err != store.ErrSnapNotFound {
+				return nil, err
+			}
+		}
+	}
+	return status, nil
+}
+
+func getThemeStatus(ctx context.Context, c *Command, user *auth.UserState, gtkThemes, iconThemes, soundThemes []string) (status themeStatusResponse, toInstall []string, err error) {
+	installedGtk, installedIcon, installedSound, err := getInstalledThemes(c.d)
+	if err != nil {
+		return themeStatusResponse{}, nil, err
+	}
+
+	theStore := getStore(c)
+	candidates := make(map[string]bool)
+	if status.GtkThemes, err = getThemeStatusForType(ctx, theStore, user, "gtk-theme-", gtkThemes, installedGtk, candidates); err != nil {
```

Comment:
It this the final way search for things providing certain class of content or do we need a TODO to pursue a dedicated store-level API?

Reply:
I think we'd want either a dedicated store API, or one that would let us query by slot definition.  The snapd level API should be high enough level not to lock in a particular design.